app/src/main/java/com/cmput301f20t21/bookfriends/ui/library/BorrowedViewModel.java

```diff
@@ -1,22 +1,96 @@
-        mData.add(new Book ("How to Squish a Cat"));
-        mData.add(new Book ("Calculus II"));
-        mData.add(new Book ("Calculus III"));
-        mData.add(new Book ("Calculus IV"));
-        mData.add(new Book ("Calculator V"));
-        mData.add(new Book ("Terminator V"));
-        return mData;
+    private RequestService requestService = RequestService.getInstance();
+    private BookService bookService = BookService.getInstance();
+    private AuthService authService = AuthService.getInstance();
+
+    private MutableLiveData<List<Book>> books;
+    private MutableLiveData<Integer> updatedPosition;
+
+    public LiveData<List<Book>> getBooks() {
+        if (books == null) {
+            books = new MutableLiveData<>();
+            fetchBooks();
+        }
+        return books;
     }
 
+    public LiveData<Integer> getUpdatedPosition() {
+        if (updatedPosition == null) {
+            updatedPosition = new MutableLiveData<>(0);
+        }
+        return updatedPosition;
+    }
+
+    private void fetchBooks() {
+        String username = authService.getCurrentUser().getUsername();
+
+        requestService.getBorrowedRequestByUsername(username).addOnSuccessListener(requestDocumentSnapshots -> {
+            // TODO: uncomment when request is implemented
+//            List<String> bookIds = requestDocumentSnapshots.getDocuments().stream()
+//                    .map(documentSnapshot -> documentSnapshot.get("bookId").toString())
+//                    .collect(Collectors.toList());
+
+            // mock book id
+            // TODO: remove this when request is implemented
+            List<String> bookIds = new ArrayList<>();
+            bookIds.add("RmJi5i1sav1B4fKQcNHP");
+            bookIds.add("asmw39EGwG2MDDPokcd7");
+            bookIds.add("x1z6o0qZbcgGBFezURsS");
+            bookIds.add("kEkNn53bBoANMSyPjJDZ");
+
+            if (bookIds.isEmpty()) {
+                return;
+            }
+
+            bookService.batchGetBooks(bookIds).addOnSuccessListener(bookDocumentsSnapshots -> {
+                List<DocumentSnapshot> documents = bookDocumentsSnapshots.getDocuments();
+                books.setValue(IntStream.range(0, documents.size()).mapToObj(i -> {
+                    DocumentSnapshot document = documents.get(i);
+                    String id = document.getId();
+                    String isbn = (String) document.get("isbn");
+                    String title = (String) document.get("title");
+                    String author = (String) document.get("author");
+                    String description = (String) document.get("description");
+                    String owner = (String) document.get("owner");
+                    String status = (String) document.get("status");
+                    String imageName = (String) document.get("imageName");
+                    Book book = new Book(id, isbn, title, author, description, owner, imageName, BOOK_STATUS.valueOf(status));
```

Comment:
This whole block could be refactored

Reply:
I can refactor this method in BookService to return a Book object for a single DocumentSnapshot
![image](https://user-images.githubusercontent.com/43459469/97526242-551d1b80-196e-11eb-952c-5c1627f54281.png)